In [1]:
import pandas as pd

In [8]:
# Load the dataset
df = pd.read_csv('/content/bank.csv')

# Display the first 5 rows
print("First 5 rows of the dataset:")
display(df.head())

First 5 rows of the dataset:


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,deposit
0,59,admin.,married,secondary,no,2343,yes,no,unknown,5,may,1042,1,-1,0,unknown,yes
1,56,admin.,married,secondary,no,45,no,no,unknown,5,may,1467,1,-1,0,unknown,yes
2,41,technician,married,secondary,no,1270,yes,no,unknown,5,may,1389,1,-1,0,unknown,yes
3,55,services,married,secondary,no,2476,yes,no,unknown,5,may,579,1,-1,0,unknown,yes
4,54,admin.,married,tertiary,no,184,no,no,unknown,5,may,673,2,-1,0,unknown,yes


In [9]:
# Check the shape of the DataFrame
print("Shape of the dataset (rows, columns):")
print(df.shape)

Shape of the dataset (rows, columns):
(11162, 17)


In [10]:
# Check the data types of the columns
print("Data types of the columns:")
print(df.dtypes)

Data types of the columns:
age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
deposit      object
dtype: object


In [11]:
# Check the column names
print("Column names:")
print(df.columns.tolist())

Column names:
['age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'deposit']


### Interpretation of Variable Types:
Based on the data types and column names, here's an initial interpretation:

*   **Numerical Variables:** `age`, `balance`, `day`, `duration`, `campaign`, `pdays`, `previous`. These are typically continuous or discrete numerical values.
*   **Categorical Variables:** `job`, `marital`, `education`, `default`, `housing`, `loan`, `contact`, `month`, `poutcome`, `y`. These variables represent categories or labels. Some might be binary (e.g., `default`, `housing`, `loan`, `y`), while others are multi-level categorical (e.g., `job`, `marital`, `education`).

Further investigation would be needed to confirm if any numerical variables are actually categorical (e.g., `day` or `month` if treated as factors rather than numerical order), or if any categorical variables have an inherent order (ordinal).

Q7. Data Cleaning

Check for missing values

Handle missing values (if any)

Check for duplicates

Interpret:

Is the dataset clean? What issues did you find?

In [12]:
# Check for missing values
print("Missing values per column:")
missing_values = df.isnull().sum()
print(missing_values)

Missing values per column:
age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
deposit      0
dtype: int64


In [13]:
# Check for duplicate rows
duplicate_rows = df.duplicated().sum()
print(f"\nNumber of duplicate rows: {duplicate_rows}")

if duplicate_rows > 0:
    print("Handling duplicate rows by dropping them...")
    df.drop_duplicates(inplace=True)
    print(f"Number of rows after dropping duplicates: {df.shape[0]}")
else:
    print("No duplicate rows found.")


Number of duplicate rows: 0
No duplicate rows found.


In [14]:
# Identify categorical columns
categorical_cols = df.select_dtypes(include='object').columns
print(f"Categorical columns to be encoded: {categorical_cols.tolist()}")

Categorical columns to be encoded: ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome', 'deposit']


In [15]:
# Apply one-hot encoding to categorical variables
df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

print("Shape of DataFrame after encoding:", df_encoded.shape)
print("First 5 rows of the encoded DataFrame:")
display(df_encoded.head())

Shape of DataFrame after encoding: (11162, 43)
First 5 rows of the encoded DataFrame:


,age,balance,day,duration,campaign,pdays,previous,job_blue-collar,job_entrepreneur,job_housemaid,...,month_jun,month_mar,month_may,month_nov,month_oct,month_sep,poutcome_other,poutcome_success,poutcome_unknown,deposit_yes
0,59,2343,5,1042,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,True
1,56,45,5,1467,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,True
2,41,1270,5,1389,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,True
3,55,2476,5,579,1,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,True
4,54,184,5,673,2,-1,0,False,False,False,...,False,False,True,False,False,False,False,False,True,True


### Interpretation: Why is encoding necessary for Decision Trees?

Encoding categorical variables into numerical format is crucial for Decision Trees (and many other machine learning algorithms) for several reasons:

1.  **Mathematical Operations:** Decision Trees, like most machine learning models, rely on mathematical operations (e.g., comparisons, splits based on thresholds) to make decisions. Categorical values (like 'admin.', 'married', 'secondary') cannot be directly compared or ordered mathematically.
2.  **Splitting Criteria:** Decision Trees build decision rules by finding optimal split points in the data. These split points are inherently numerical. For a categorical feature, the tree needs to be able to evaluate different categories to determine the best way to divide the data. Encoding transforms these categories into a numerical representation that can be used in these splitting criteria.
3.  **Preventing Ordinality Misinterpretation (if using One-Hot Encoding):** While some encoding methods (like Label Encoding) assign a single integer to each category, this can imply an ordinal relationship where none exists (e.g., `job` = 'admin' might become 0, 'blue-collar' 1, 'entrepreneur' 2, implying 2 > 1 > 0). One-Hot Encoding creates separate binary columns for each category, preventing the model from falsely assuming such an order. Decision Trees can handle ordinal features naturally if they are encoded numerically, but for nominal features (where no inherent order exists), one-hot encoding is generally preferred to avoid misleading the algorithm.

In essence, encoding converts qualitative information into a quantitative format that the Decision Tree algorithm can understand and process to build its decision-making logic.

### Interpretation: Is the dataset clean? What issues did you find?

Based on the checks performed:

*   **Missing Values:** All columns show `0` missing values. This indicates that the dataset is complete and does not require imputation or removal of rows/columns due to missing data.
*   **Duplicate Rows:** The check for duplicates found `0` duplicate rows. This means every row in the dataset is unique, and no rows needed to be removed.

**Conclusion:** The dataset appears to be relatively clean in terms of missing values and duplicate entries. No immediate data cleaning issues related to these aspects were identified. Further steps might involve examining outliers, correcting inconsistent data entries, or handling data types more specifically if required for subsequent analysis.

Q8. Data Preprocessing

Convert categorical variables into numerical

Mention which encoding method you used

Interpret:

Why is encoding necessary for Decision Trees?

In [16]:
from sklearn.preprocessing import StandardScaler

# Identify numerical columns (excluding the encoded categorical columns and 'deposit' which is the target)
# Numerical columns are those that are not boolean (from one-hot encoding) and are not the target variable
numerical_cols = df_encoded.select_dtypes(include=['int64', 'float64']).columns.tolist()

print(f"Numerical columns to be scaled: {numerical_cols}")

Numerical columns to be scaled: ['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']


In [17]:
# Initialize the StandardScaler
scaler = StandardScaler()

# Apply scaling to the numerical columns
df_encoded[numerical_cols] = scaler.fit_transform(df_encoded[numerical_cols])

print("Shape of DataFrame after scaling numerical features:", df_encoded.shape)
print("First 5 rows of the scaled DataFrame:")
display(df_encoded.head())

Shape of DataFrame after scaling numerical features: (11162, 43)
First 5 rows of the scaled DataFrame:


,age,balance,day,duration,campaign,pdays,previous,job_blue-collar,job_entrepreneur,job_housemaid,...,month_jun,month_mar,month_may,month_nov,month_oct,month_sep,poutcome_other,poutcome_success,poutcome_unknown,deposit_yes
0,1.491505,0.252525,-1.265746,1.930226,-0.554168,-0.481184,-0.36326,False,False,False,...,False,False,True,False,False,False,False,False,True,True
1,1.239676,-0.459974,-1.265746,3.154612,-0.554168,-0.481184,-0.36326,False,False,False,...,False,False,True,False,False,False,False,False,True,True
2,-0.019470,-0.080160,-1.265746,2.929901,-0.554168,-0.481184,-0.36326,False,False,False,...,False,False,True,False,False,False,False,False,True,True
3,1.155733,0.293762,-1.265746,0.596366,-0.554168,-0.481184,-0.36326,False,False,False,...,False,False,True,False,False,False,False,False,True,True
4,1.071790,-0.416876,-1.265746,0.867171,-0.186785,-0.481184,-0.36326,False,False,False,...,False,False,True,False,False,False,False,False,True,True


### Interpretation: Why is Feature Scaling Necessary?

Feature scaling, particularly using methods like Standardization (as performed above with `StandardScaler`), is crucial for many machine learning algorithms, although its importance varies:

1.  **Algorithms Sensitive to Feature Scales:** Algorithms that use distance calculations (e.g., K-Nearest Neighbors, Support Vector Machines, K-Means Clustering) or gradient descent optimization (e.g., Logistic Regression, Neural Networks) are highly sensitive to the scale of input features. If features have vastly different ranges, features with larger values can dominate the distance calculation or gradient updates, leading to suboptimal model performance.
2.  **Gradient Descent Optimization:** In algorithms that use gradient descent, unscaled features can lead to an elongated and narrow cost function surface. This makes it difficult for the optimization algorithm to find the global minimum efficiently, potentially causing it to oscillate or take longer to converge.
3.  **Regularization:** Regularization techniques (L1, L2) penalize large coefficients. If features are not scaled, coefficients for features with smaller scales might be unduly penalized or, conversely, features with larger scales might not be sufficiently regularized, leading to biased models.

### Why it might be less critical for Decision Trees:

It's important to note that **Decision Trees and tree-based ensemble methods (like Random Forests and Gradient Boosting) are generally not sensitive to feature scaling.** This is because:

*   **Splitting Criteria:** Decision trees make splits based on individual feature values and thresholds (e.g., `feature_A > 10`). The exact scale of the feature doesn't affect which feature is chosen for a split or where the split point is, as long as the relative order of values is maintained. They don't use distance metrics or gradient descent in their core splitting mechanism.

However, it's a good practice to scale features during preprocessing, especially if you plan to experiment with different models later in the machine learning pipeline, some of which *will* benefit significantly from scaled data.

## Q9. Feature Selection & Splitting

In [18]:
# Separate features (X) and target (y)
# The target variable 'deposit_yes' was created during one-hot encoding.
X = df_encoded.drop('deposit_yes', axis=1)
y = df_encoded['deposit_yes']

print("Shape of features (X):", X.shape)
print("Shape of target (y):", y.shape)

Shape of features (X): (11162, 42)
Shape of target (y): (11162,)


In [19]:
from sklearn.model_selection import train_test_split

# Perform an 80-20 train-test split
# stratify=y ensures that both training and testing sets have a similar proportion of target classes
# random_state for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

print("\nClass distribution in original target (y):\n", y.value_counts(normalize=True))
print("\nClass distribution in training target (y_train):\n", y_train.value_counts(normalize=True))
print("\nClass distribution in testing target (y_test):\n", y_test.value_counts(normalize=True))

X_train shape: (8929, 42)
X_test shape: (2233, 42)
y_train shape: (8929,)
y_test shape: (2233,)

Class distribution in original target (y):
 deposit_yes
False    0.52616
True     0.47384
Name: proportion, dtype: float64

Class distribution in training target (y_train):
 deposit_yes
False    0.526151
True     0.473849
Name: proportion, dtype: float64

Class distribution in testing target (y_test):
 deposit_yes
False    0.526198
True     0.473802
Name: proportion, dtype: float64


### Interpretation: Feature Selection and Train-Test Split

1.  **Feature and Target Separation:**
    *   We successfully separated the `df_encoded` DataFrame into `X` (features) and `y` (target variable). The `deposit_yes` column, which indicates whether a customer subscribed to a term deposit, was designated as the target (`y`). All other columns, representing the customer's characteristics, form the feature set (`X`).

2.  **Train-Test Split:**
    *   The dataset was split into training and testing sets using an 80-20 ratio (80% for training, 20% for testing). This is a standard practice to evaluate the model's performance on unseen data, preventing overfitting.
    *   `random_state=42` ensures that the split is reproducible, meaning you'll get the same split every time you run the code.
    *   `stratify=y` is particularly important for classification tasks with imbalanced datasets. It ensures that the proportion of target classes (i.e., 'yes' vs. 'no' for deposit) is roughly the same in both the training and testing sets as it was in the original dataset. This helps in building a model that is representative of the actual class distribution and ensures a fair evaluation.

Now, the data is ready for model training using `X_train` and `y_train`, and subsequent evaluation using `X_test` and `y_test`.

## Q10. Model Training and Evaluation

In [20]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report, confusion_matrix

# Initialize the Decision Tree Classifier
# Set random_state for reproducibility
dt_classifier = DecisionTreeClassifier(random_state=42)

# Train the model on the training data
print("Training Decision Tree Classifier...")
dt_classifier.fit(X_train, y_train)
print("Training complete.")

# Make predictions on the test set
y_pred = dt_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"\nAccuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Training Decision Tree Classifier...
Training complete.

Accuracy: 0.7936
Precision: 0.7878
Recall: 0.7722
F1-Score: 0.7800

Classification Report:
              precision    recall  f1-score   support

       False       0.80      0.81      0.81      1175
        True       0.79      0.77      0.78      1058

    accuracy                           0.79      2233
   macro avg       0.79      0.79      0.79      2233
weighted avg       0.79      0.79      0.79      2233


Confusion Matrix:
[[955 220]
 [241 817]]


### Interpretation: Decision Tree Model Performance

The Decision Tree Classifier was trained on the preprocessed data and evaluated on the test set. Here's an interpretation of the performance metrics:

*   **Accuracy:** This measures the proportion of correctly classified instances out of the total instances. An accuracy of **79.36%** means that the model correctly predicted the outcome (whether a customer subscribes to a term deposit or not) for 79.36% of the test cases.

*   **Precision:** For the 'yes' class (customers who subscribe), precision measures the proportion of positive identifications that were actually correct. A precision of **78.78%** means that when the model predicts a customer will subscribe, it is correct 78.78% of the time. High precision indicates a low false positive rate.

*   **Recall (Sensitivity):** For the 'yes' class, recall measures the proportion of actual positives that were identified correctly. A recall of **77.22%** means that the model correctly identified 77.22% of all customers who *actually did* subscribe. High recall indicates a low false negative rate.

*   **F1-Score:** This is the harmonic mean of precision and recall, providing a single metric that balances both. An F1-score of **78.00%** indicates a reasonable balance between precision and recall for the 'yes' class.

*   **Classification Report:** This provides a more detailed breakdown of precision, recall, and F1-score for both classes ('no' and 'yes'), along with support (the number of actual occurrences of each class in the test set).

*   **Confusion Matrix:**
    *   **True Positives (TP):** The number of customers correctly predicted as subscribing (actual 'yes', predicted 'yes'). In this case, **817**.
    *   **True Negatives (TN):** The number of customers correctly predicted as not subscribing (actual 'no', predicted 'no'). In this case, **955**.
    *   **False Positives (FP):** The number of customers incorrectly predicted as subscribing (actual 'no', predicted 'yes'). These are Type I errors. In this case, **220**.
    *   **False Negatives (FN):** The number of customers incorrectly predicted as not subscribing (actual 'yes', predicted 'no'). These are Type II errors. In this case, **241**.

**Overall Assessment:** The Decision Tree model shows a moderate to good performance with an overall accuracy of 79.36%. The precision (78.78%) and recall (77.22%) for predicting 'yes' subscriptions are fairly balanced, indicating that the model is reasonably effective at identifying potential subscribers while also keeping false positives in check. The F1-score of 78.00% reinforces this balance. Depending on the business objective (e.g., minimizing false positives vs. false negatives), one metric might be more critical than others, but overall, this model provides a solid baseline performance for predicting term deposit subscriptions.